Add --serverfingerprint option

lukechilds opened the pull request:
Resolves #6084 

Needs a bit more work but it's fully functional and testable now.

If you want to test against my server you can run:

```
$ ./run_electrum --oneserver --server bitcoin.lukechilds.co:50002:s --serverfingerprint ED1FE28B21A5943A8810747B2A6DDFC0E3089115D6205125E5C4D3EE8E126F80
```

This should connect as normal.

If you run:

```
$ ./run_electrum --oneserver --server bitcoin.lukechilds.co:50002:s --serverfingerprint DEADBEEF
```

Then an `ErrorSSLCertFingerprintMismatch` error should be thrown.

Two questions:

1. `ErrorSSLCertFingerprintMismatch` is being caught somewhere so it doesn't kill the app, the connection keeps being retried. It's not entirely clear to me where this is getting handled. Instead I want the error handler to exit the application if it encounters an `ErrorSSLCertFingerprintMismatch`. Where should I add this logic?

2. I need to display an error dialog to GUI users before exit, informing them of the fingerprint mismatch. Is it ok to do Qt stuff directly inside `interface.py`? It seems like maybe that's not the right place. What's the best way to structure that?

ROSAISS81 commented:
rosaiselaperez@gmail.com rosa Isela perez martinez
te  soy sincera no se lo  puedes hacer   tu te lo agradecería tanto y una
pregunta  paraqué es o que es esto no lo se en verdad lo siento pero es la
verdad

El jue., 16 de abr. de 2020 a la(s) 03:54, Luke Childs (
notifications@github.com) escribió:

> Resolves #6084 <https://github.com/spesmilo/electrum/issues/6084>
>
> Needs a bit more work but it's fully functional and testable now.
>
> If you want to test against my server you can run:
>
> $ ./run_electrum --oneserver --server bitcoin.lukechilds.co:50002:s --fingerprint ED1FE28B21A5943A8810747B2A6DDFC0E3089115D6205125E5C4D3EE8E126F80
>
> This should connect as normal.
>
> If you run:
>
> $ ./run_electrum --oneserver --server bitcoin.lukechilds.co:50002:s --fingerprint DEADBEEF
>
> Then an ErrorSSLCertFingerprintMismatch error should be thrown.
>
> Two questions:
>
>    1.
>
>    ErrorSSLCertFingerprintMismatch is being caught somewhere so it
>    doesn't kill the app, the connection keeps being retried. It's not entirely
>    clear to me where this is getting handled. Instead I want the error handler
>    to exit the application if it encounters an
>    ErrorSSLCertFingerprintMismatch. Where should I add this logic?
>    2.
>
>    I need to display an error dialog to GUI users before exit, informing
>    them of the fingerprint mismatch. Is it ok to do Qt stuff directly inside
>    interface.py? It seems like maybe that's not the right place.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/spesmilo/electrum/pull/6094
> Commit Summary
>
>    - Add electrum_data to .gitignore
>    - Add --fingerprint option
>    - Simplify conditional checks
>    - Improve warning wording
>    - Throw error instead of logging and returning
>
> File Changes
>
>    - *M* .gitignore
>    <https://github.com/spesmilo/electrum/pull/6094/files#diff-a084b794bc0759e7a6b77810e01874f2>
>    (1)
>    - *M* electrum/commands.py
>    <https://github.com/spesmilo/electrum/pull/6094/files#diff-b38f023e655dd0d1f65598f2b2ca204b>
>    (1)
>    - *M* electrum/interface.py
>    <https://github.com/spesmilo/electrum/pull/6094/files#diff-467d1730bca804f9d459a7e2d0036a89>
>    (11)
>
> Patch Links:
>
>    - https://github.com/spesmilo/electrum/pull/6094.patch
>    - https://github.com/spesmilo/electrum/pull/6094.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spesmilo/electrum/pull/6094>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMAHBJNDNDL2JWEPUPABALTRM3PYBANCNFSM4MJQHHVA>
> .
>

SomberNight commented on electrum/commands.py:
```diff
@@ -1258,6 +1258,7 @@ def subparser_call(self, parser, namespace, values, option_string=None):
 
 
 def add_network_options(parser):
+    parser.add_argument("-f", "--fingerprint", dest="fingerprint", default=None, help="only allow connecting to servers with a matching SSL certificate SHA256 fingerprint")
```
`fingerprint` sounds a bit too generic, it might become ambiguous in the future.
maybe `sslfingerprint` or `serverfingerprint` or `sslcerthash`? well not sure.

lukechilds replied:
I agree, I like `serverfingerprint`.

lukechilds replied:
Resolved by https://github.com/spesmilo/electrum/pull/6094/commits/0fcde1306197927de3793326aba840df2ad7c80a

SomberNight commented on electrum/interface.py:
```diff
@@ -404,6 +406,9 @@ async def wrapper_func(self: 'Interface', *args, **kwargs):
     async def run(self):
+        expected_fingerprint = self.network.config.get("fingerprint")
+        if expected_fingerprint and not await self.verify_server_certificate(expected_fingerprint):
+            raise ErrorSSLCertFingerprintMismatch('Refusing to connect to server due to SSL certificate fingerprint mismatch')
```
if `self.protocol` is not "s", you should not run ssl-related code, e.g. `self.get_certificate()`.
Instead, if protocol is not "s" but fingerprint is provided, you could just fail with a distinct error.

lukechilds replied:
Implemented the check in https://github.com/spesmilo/electrum/pull/6094/commits/6d7698ec0b5d9078fb8c2e0c9dd5c334925e7533 but having the same issue I mentioned in the original comment with the error being caught somewhere instead of causing the application to exit.

lukechilds replied:
This is now resolved after refactoring the fingerprint check to be next to the other certificate related logic.

lukechilds commented on electrum/interface.py:
```diff
@@ -404,6 +406,9 @@ async def wrapper_func(self: 'Interface', *args, **kwargs):
     async def run(self):
+        expected_fingerprint = self.network.config.get("fingerprint")
+        if expected_fingerprint and not await self.verify_server_certificate(expected_fingerprint):
+            raise ErrorSSLCertFingerprintMismatch('Refusing to connect to server due to SSL certificate fingerprint mismatch')
```
It wasn't deliberate, this should be limited to just the main server.

lukechilds replied:
Resolved by https://github.com/spesmilo/electrum/pull/6094/commits/e2a8b347dc760693ced2ff0c6923eb4fa06b198d

lukechilds commented on electrum/interface.py:
```diff
@@ -471,6 +476,12 @@ async def get_certificate(self):
             return None
 
+    async def verify_server_certificate(self, expected_fingerprint):
+        certificate = await self.get_certificate()
```
What would you suggest here?

Add a simple cache to `get_certificate()` so it only fetches the cert the first time and returns it in future calls?

lukechilds replied:
Oh sorry, I think misunderstood what you meant. This problem is mitigated by  https://github.com/spesmilo/electrum/pull/6094/commits/e2a8b347dc760693ced2ff0c6923eb4fa06b198d, right?

SomberNight replied:
Well it's somewhat mitigated, yes. However it's still a mildly annoying that we do extra network round trips for the main server if the feature is enabled. Still, as it only happens if a fingerprint is provided, it's fine.
--- But, what's not fine is that due to the implementation, and this extra round trip, these checks do not actually guarantee any extra security. :P
The evil ISP can just look at the code, and realise it's enough to pass through the real certificate on the first round-trip, which we check, and then present their own evil cert (which will get pinned) on the second round-trip.

SomberNight replied:
So, my suggestion is this:
- for simplicity, disallow CA-signed certs when a fingerprint is provided (raise distinct error)
- change the API of `verify_server_certificate` to take the cert itself as input
- call `verify_server_certificate` from `interface._is_saved_ssl_cert_available`, and also from `interface.save_certificate` (make sure the latter does not save the cert if the fingerprint mismatches, ofc)

lukechilds replied:
Ah yes, my intention was to validate the actual certificate used in the current connection, not make a specific request to get one. Doh!

To allow this to work with CA certs could I not just use [`getpeercert`](https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.getpeercert) on the actual socket that is being used for the main connection, and then check the fingerprint against that?

That way it'll just be one check in one place for all scenarios. CA/self signed/initial connection/future pinned connection, it'll all be checked the same way.

SomberNight replied:
You could do that, but with how the code is structured now (on master), by the time you have the actual socket, some form of cert pinning has already happened. (i.e. you might end up having pinned the wrong cert)

lukechilds replied:
This should be resolved in https://github.com/spesmilo/electrum/pull/6094/commits/54a26d6d595c117c121094c1083641bae581e410 and https://github.com/spesmilo/electrum/pull/6094/commits/416a41e9bf420239944a5882d40ff8830be52409

SomberNight commented:
> I need to display an error dialog to GUI users before exit, informing them of the fingerprint mismatch. Is it ok to do Qt stuff directly inside interface.py? It seems like maybe that's not the right place. What's the best way to structure that?

Qt GUI elements must only be created/modified from the GUI thread.
This code in interface all runs on the asyncio event loop, which runs on a separate thread.
Search for e.g. `util.trigger_callback('blockchain_updated')`, and see
https://github.com/spesmilo/electrum/blob/cd199390e2d6376cdf9458c4a14906c564beea1c/electrum/gui/qt/main_window.py#L265
https://github.com/spesmilo/electrum/blob/cd199390e2d6376cdf9458c4a14906c564beea1c/electrum/gui/qt/main_window.py#L407

> ErrorSSLCertFingerprintMismatch is being caught somewhere so it doesn't kill the app, the connection keeps being retried. It's not entirely clear to me where this is getting handled. Instead I want the error handler to exit the application if it encounters an ErrorSSLCertFingerprintMismatch. Where should I add this logic?

I am not sure if we should exit the application.
The exception is getting caught btw in the decorators of `interface.run`.

Maybe instead of exiting, it's good enough to notify the user and keep retrying.
Anyone who reads logs will see the issue; so we should just make sure GUI users get some feedback too.
Also note that if you display some feedback in the GUI and you also want to kill the application, you should make sure the user has time to read the feedback.
Another point is how to show feedback to GUI users: if we show a popup, we should probably kill the application, as spamming popups would be horrible (if we keep retrying). Alternatively if we don't kill the application, we could maybe show OS notifications (although even spamming those is kinda bad).

So, maybe what we could do is:
- for Qt GUI users, show popup, wait for user to click OK, and then quit app
- for Kivy GUI users, nothing atm, as they are not able to use this feature for now
- others, we assume they read logs, and we just keep retrying

lukechilds commented:
Thanks for clarifying that, that makes a lot of sense!

It's late here now but I'll crack on with this tomorrow.

lukechilds commented:
I think I've addressed all the points in your feedback.

This is what a GUI user sees if they try to connect to a server with an invalid fingerprint:

<img width="952" alt="Screenshot 2020-04-19 at 14 11 35" src="https://user-images.githubusercontent.com/2123375/79681864-f8ff6f00-8247-11ea-8cce-1b0bac6161a9.png">

Let me know if the wording is ok or if you want me to change it.

SomberNight commented on electrum/gui/qt/main_window.py:
```diff
@@ -3118,3 +3122,14 @@ def save_transaction_into_wallet(self, tx: Transaction):
+            return
+        self.showing_cert_mismatch_error = True
+        self.show_critical(title=_("Certificate mismatch"),
+                           msg="\n\n" +
```
Why the leading newlines? It looks a bit weird on both Ubuntu and Windows.
I see from your picture that on macOS it does not show the title for the popup -- weird. In any case I think we should rm these newlines, at least for uniformity with other popups.

lukechilds replied:
Ah, I was wondering why that didn't show!

The leading newlines were original because I had "Certificate mismatch" at the beginning of `msg` because it wasn't showing in `title`. When I removed the duplicate "Certificate mismatch" I must've forgotten to remove the leading newlines along with it.

SomberNight commented on electrum/interface.py:
```diff
@@ -350,6 +353,8 @@ async def is_server_ca_signed(self, ca_ssl_context):
     async def _try_saving_ssl_cert_for_first_time(self, ca_ssl_context):
         ca_signed = await self.is_server_ca_signed(ca_ssl_context)
         if ca_signed:
+            if self.network.config.get("serverfingerprint"):
+                raise InvalidOptionCombination("cannot use --serverfingerprint with CA signed servers")
```
This should also test for `is_main_server`.

SomberNight commented on electrum/interface.py:
```diff
@@ -362,6 +367,8 @@ def _is_saved_ssl_cert_available(self):
         with open(self.cert_path, 'r') as f:
             contents = f.read()
         if contents == '':  # CA signed
+            if self.network.config.get("serverfingerprint"):
+                raise InvalidOptionCombination("cannot use --serverfingerprint with CA signed servers")
```
As above, this should also test for `is_main_server`.

How about introducing a new method, `interface._get_expected_fingerprint` or similar, which returns
`self.network.config.get("serverfingerprint")` for the main server, and None otherwise? That way, the `is_main_server` tests don't have to be duplicated.

SomberNight commented on electrum/commands.py:
```diff
@@ -1263,6 +1263,7 @@ def subparser_call(self, parser, namespace, values, option_string=None):
 
 
 def add_network_options(parser):
+    parser.add_argument("-f", "--serverfingerprint", dest="serverfingerprint", default=None, help="only allow connecting to servers with a matching SSL certificate SHA256 fingerprint")
```
Maybe we could also say here how to calc the fingerprint for yourself.
`To calculate this yourself: '$ openssl x509 -noout -fingerprint -sha256 -inform pem -in mycertfile.crt'. Enter as 64 hex chars.`

lukechilds commented:
Latest feedback addressed, let me know if it's ok.

lukechilds commented:
@SomberNight just pinging you to make sure you've seen this. If you have and you just haven't gotten around to it yet, no worries!
